pyface/base_application.py

```diff
@@ -0,0 +1,286 @@
+
+    #: Upon successful completion of the stop method.
+    stopped = Event(ApplicationEvent)
+
+    # Protected interface -----------------------------------------------------
+
+    #: Flag if the exiting of the application was explicitely requested by user
+    # An 'explicit' exit is when the 'exit' method is called.
+    # An 'implicit' exit is when the user closes the last open window.
+    _explicit_exit = Bool(False)
+
+    # Application lifecycle methods -------------------------------------------
+
+    def __init__(self, **traits):
+        """ Initialize the application """
+        super(BaseApplication, self).__init__(**traits)
+
+        # perform basic initialization before we can do anything
+        self.init()
+
+    def init(self):
+        """ Perform basic application configuration
+
+        In a full application, this includes setting up the application's home
+        directory, setting up basic logging, handling command-line options and
+        setting user preferences.
+        """
+        logger.info('---- Application configuration ----')
+
+        # install our own exception hook for unhandled exceptions
+        sys.excepthook = self._excepthook
+        logger.debug('Exception hook installed')
+
+        # make sure we have a home directory
+        self._initialize_application_home()
+
+    def start(self):
+        """ Start the application, setting up things that are required
+
+        Subclasses should open at least one ApplicationWindow or subclass in
+        their start method, and should call the superclass start() method
+        before doing any work themselves.
+        """
+        logger.info('---- Application starting ----')
+        return True
+
+    def stop(self):
+        """ Stop the application, cleanly releasing resources if possible. """
+        logger.info('---- Application stopping ----')
+        return True
+
+    def run(self):
+        """ Run the application """
+
+        # Start up the application.
+        self.starting = ApplicationEvent(application=self,
+                                         event_type='starting')
+        started = self.start()
+        if started:
+            logger.info('---- Application started ----')
+            self.started = ApplicationEvent(application=self,
+                                            event_type='started')
+
+            self._run()
+
+            # We have finished running, so shut the application down.
+            self.stopping = ApplicationEvent(application=self,
+                                             event_type='stopping')
+            stopped = self.stop()
+            if stopped:
+                self.stopped = ApplicationEvent(application=self,
+                                                event_type='stopped')
+                logger.info('---- Application stopped ----')
+
+                # exit normally
+                sys.exit()
+
+        # exit with an error
+        sys.exit(1)
```

Comment:
What is the point of these calls to `sys.exit`, as opposed to exiting with an exception?